Floating platform (#2485)
jpmorganchase/salt-ds: packages/core/stories/floating-platform/floating-platform.doc.stories.mdx

import { ArgsTable, Canvas, Meta, Story } from "@storybook/addon-docs";
import { Button, Text, Tooltip } from "@salt-ds/core";
import {
  CharacteristicUsage,
  CSSClassTable,
} from "css-variable-docgen-components";
import HelpAndSupport from "docs/blocks/help-and-support.doc.mdx";

<Meta
  title="Documentation/Core/Floating Platform"
  component={Tooltip}
  parameters={{
    viewMode: "docs",
  }}
/>

## Custom Floating UI Platform

Note: The following section should not be necessary for the majority of users and is only fot specialised use-cases of floating components

### Background

> Certain components within the design system use 'floating' . These are elements which are anchored to another component, but appear outside of the normal flow of the document, and re-posiiton themselves according to factors such as proximaty to the ege of the document. Salt uses [Floating UI](https://floating-ui.com) to implement these aspect of our components.

Reviewer: ```suggestion
Certain components within the design system use 'floating' . These are elements which are anchored to another component, but appear outside of the normal flow of the document, and re-position themselves according to factors such as proximity to the edge of the document. Salt uses [Floating UI](https://floating-ui.com) to implement these aspect of our components.
```

> In certain specialised use-cases consumers may want to customise the behaviour of floating-ui. For example when dealing with multi-window applications, where floating components should spawn in a new window and re-position with proximaty to the user's screen.

Reviewer: ```suggestion
In certain specialized use-cases consumers may want to customize the behavior of floating-ui. For example when dealing with multi-window applications, where floating components should spawn in a new window and re-position with proximity to the user's screen.
```

### Custom Platforms

> To facilitate this advanced usage consumer's can pass a custom platform, based on [Floating UI's Platform feature](https://floating-ui.com/docs/platform)

Reviewer: ```suggestion
To facilitate this advanced usage consumers can pass a custom platform, based on [Floating UI's Platform feature](https://floating-ui.com/docs/platform)
```

You can provide a custom Floating UI platform to be used in these components by importing `<FloatingPlatformProvider>` and passing a custom [Floating UI Platform](https://floating-ui.com/docs/platform) as the `platform` prop. This will then be used within the floating component.

### Custom Middleware

> Using the FloatingPlatformProvider you can also pass custom [Floating UI Middleware](https://floating-ui.com/docs/middleware) to be applied on top of the defaults for tweaking the position of floating components using the `middleware` prop.

Reviewer: ```suggestion
Using the `FloatingPlatformProvider` you can also pass custom [Floating UI Middleware](https://floating-ui.com/docs/middleware) to be applied on top of the defaults for tweaking the position of floating components using the `middleware` prop.
```

This custom component will probably need to use [FloatingPortal](https://floating-ui.com/docs/FloatingPortal) and also add a `SaltProvider` at it's root.

```
import { Salt } from "@salt-ds/core";
import { FloatingComponentProvider } from "@salt-ds/window";
import { FloatingPortal } from "@floating-ui/react";
```

```
const FloatingComponent = forwardRef(
  (
    props: CombinedFloatingComponentProps,
    ref: ForwardedRef<HTMLDivElement>
  ) => {
    const { open, disabled, ...rest } = props;
    return open && !disabled ? (
      <FloatingPortal>
        <SaltProvider>
          <div {...rest} ref={ref} />
        </SaltProvider>
      </FloatingPortal>
    ) : null;
  }
);

const CustomFloatingComponentTooltip = ({...props}) => {
    return (
        <FloatingComponentProvider
            Component={FloatingComponent}
        >
            <Tooltip {...props} />
        </FloatingComponentProvider>
    )
}
```

### Custom Platform Example

> The following example shows the two providers in action. The application is rendered in an iframe, representing a windowed applicationinside a global coordinate space (e.g. a users desktop).

Reviewer: ```suggestion
The following example shows the two providers in action. The application is rendered in an iframe, representing a windowed application inside a global coordinate space (e.g. a users desktop).
```
